aliasmgr.go

```diff
@@ -0,0 +1,446 @@
+		}
+	}
+
+	return nil
+}
+
+// putPeerAlias stores the peer's alias SCID once we learn of it in the
+// funding_locked message.
+func (a *aliasMgr) putPeerAlias(chanID lnwire.ChannelID,
+	alias lnwire.ShortChannelID) error {
+
+	return kvdb.Update(a.backend, func(tx kvdb.RwTx) error {
+		bucket, err := tx.CreateTopLevelBucket(invoiceAliasBucket)
+		if err != nil {
+			return err
+		}
+
+		var scratch [8]byte
+		byteOrder.PutUint64(scratch[:], alias.ToUint64())
+		return bucket.Put(chanID[:], scratch[:])
+	}, func() {})
+}
+
+// getPeerAlias retrieves a peer's alias SCID by the channel's ChanID.
+func (a *aliasMgr) getPeerAlias(chanID lnwire.ChannelID) (
+	lnwire.ShortChannelID, error) {
+
+	var alias lnwire.ShortChannelID
+
+	err := kvdb.Update(a.backend, func(tx kvdb.RwTx) error {
+		bucket, err := tx.CreateTopLevelBucket(invoiceAliasBucket)
+		if err != nil {
+			return err
+		}
+
+		aliasBytes := bucket.Get(chanID[:])
+		if aliasBytes == nil {
+			return nil
+		}
+
+		alias = lnwire.NewShortChanIDFromInt(
+			byteOrder.Uint64(aliasBytes),
+		)
+		return nil
+	}, func() {})
+
+	if alias == hop.Source {
+		return alias, errNoPeerAlias
+	}
+
+	return alias, err
+}
+
+// requestAlias returns a new ALIAS ShortChannelID to the caller by allocating
+// the next un-allocated ShortChannelID. The starting ShortChannelID is
+// 10000:0:0 and the ending ShortChannelID is 262143:16777215:65535. This gives
+// roughly 2^58 possible ALIAS ShortChannelIDs which ensures this space won't
+// get exhausted.
+func (a *aliasMgr) requestAlias() (lnwire.ShortChannelID, error) {
+	var nextAlias lnwire.ShortChannelID
+
+	err := kvdb.Update(a.backend, func(tx kvdb.RwTx) error {
+		bucket, err := tx.CreateTopLevelBucket(aliasAllocBucket)
+		if err != nil {
+			return err
+		}
+
+		lastBytes := bucket.Get(lastAliasKey)
+		if lastBytes == nil {
+			// If the key does not exist, then we can write the
+			// startingAlias to it.
+			nextAlias = startingAlias
+
+			var scratch [8]byte
+			byteOrder.PutUint64(scratch[:], nextAlias.ToUint64())
+			return bucket.Put(lastAliasKey, scratch[:])
+		}
+
+		// Otherwise the key does exist so we can convert the retrieved
+		// lastAlias to a ShortChannelID and use it to assign the next
+		// ShortChannelID. This next ShortChannelID will then be
+		// persisted in the database.
+		lastScid := lnwire.NewShortChanIDFromInt(
+			byteOrder.Uint64(lastBytes),
+		)
+		nextAlias = getNextScid(lastScid)
+
+		var scratch [8]byte
+		byteOrder.PutUint64(scratch[:], nextAlias.ToUint64())
+		return bucket.Put(lastAliasKey, scratch[:])
+	}, func() {})
+	if err != nil {
+		return nextAlias, err
+	}
+
+	return nextAlias, nil
+}
+
+// getNextScid is a utility function that returns the next SCID for a given
+// alias SCID. The BlockHeight ranges from [10000, 262143], the TxIndex ranges
+// from [1, 16777215], and the TxPosition ranges from [1, 65535].
+func getNextScid(last lnwire.ShortChannelID) lnwire.ShortChannelID {
```

Comment:
tradeoff is having to store every single allocated one

Reply:
There may be some way to generate these deterministically from a "seed" and then just storing the seed and "height" in the database. That way it's not the same for everybody. Though I haven't thought too much about it

Reply:
I think we could just use a random starting alias determined once that isn't close to the end of the range. I have no idea what that should be though. I think it would leak info though:
- A chooses starting alias in the blockheight range [10000, 65000]
- B sees A has an alias of 15000, concluding that A hasn't used too many ailases
OR
- B sees A has an alias of 45000, and is less sure whether A is new or not